File: src/main/java/nessiesson/usefulmod/ShulkerBoxDisplay.java
package nessiesson.usefulmod;

import net.minecraft.block.BlockShulkerBox;
import net.minecraft.client.Minecraft;
import net.minecraft.client.gui.Gui;
import net.minecraft.client.gui.GuiScreen;
import net.minecraft.client.renderer.GlStateManager;
import net.minecraft.client.renderer.RenderHelper;
import net.minecraft.client.renderer.RenderItem;
import net.minecraft.client.resources.I18n;
import net.minecraft.inventory.ItemStackHelper;
import net.minecraft.item.EnumDyeColor;
import net.minecraft.item.ItemBlock;
import net.minecraft.item.ItemShulkerBox;
import net.minecraft.item.ItemStack;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.util.NonNullList;
import net.minecraft.util.ResourceLocation;

import java.util.List;

// Modified version of the shulkerbox display from Vazkii's Quark Forgemod.
public class ShulkerBoxDisplay {
	private static final ResourceLocation WIDGET_RESOURCE = new ResourceLocation("usefulmod", "textures/shulker_widget.png");

	public static void handleShulkerBoxDisplayRenderer(ItemStack stack, int x, int y, Gui gui) {
		if (!LiteModUsefulMod.config.shulkerBoxDisplay) {
			return;
		}

		if (stack != null && stack.getItem() instanceof ItemShulkerBox && stack.hasTagCompound() && GuiScreen.isShiftKeyDown()) {
			NBTTagCompound cmp = getCompoundOrNull(stack);
			if (cmp != null && cmp.hasKey("Items", 9)) {
				final int texWidth = 172;
				final int texHeight = 64;

				final int dy = y - texHeight - 18;

				GlStateManager.pushMatrix();
				RenderHelper.enableGUIStandardItemLighting();
				GlStateManager.enableRescaleNormal();
				GlStateManager.translate(0F, 0F, 700F);

				final Minecraft mc = Minecraft.getMinecraft();
				mc.getTextureManager().bindTexture(WIDGET_RESOURCE);

				final EnumDyeColor dye = ((BlockShulkerBox) ((ItemBlock) stack.getItem()).getBlock()).getColor();
				final float[] colours = dye.getColorComponentValues();
				GlStateManager.color(colours[0], colours[1], colours[2]);

				gui.drawTexturedModalRect(x, dy, 0, 0, texWidth, texHeight);

				NonNullList<ItemStack> itemList = NonNullList.withSize(27, ItemStack.EMPTY);
				ItemStackHelper.loadAllItems(cmp, itemList);

				RenderItem render = mc.getRenderItem();

				GlStateManager.enableDepth();
				int i = 0;
				for (ItemStack itemstack : itemList) {
					if (!itemstack.isEmpty()) {
						final int xp = x + 6 + (i % 9) * 18;
						final int yp = dy + 6 + (i / 9) * 18;

						render.renderItemAndEffectIntoGUI(itemstack, xp, yp);
						render.renderItemOverlays(mc.fontRenderer, itemstack, xp, yp);
					}

					i++;
				}

				GlStateManager.disableDepth();
				GlStateManager.popMatrix();
			}
		}
	}

	public static void addShulkerBoxTooltip(ItemStack stack, List<String> tooltip) {
		if (!LiteModUsefulMod.config.shulkerBoxDisplay) {
			return;
		}

		if (!stack.isEmpty() && stack.getItem() instanceof ItemShulkerBox && stack.hasTagCompound()) {
			final NBTTagCompound cmp = getCompoundOrNull(stack);
			if (cmp != null && cmp.hasKey("Items", 9)) {

				if (!GuiScreen.isShiftKeyDown()) {
					tooltip.add(1, I18n.format("usefulmod.ui.mention_shift"));
				}
			}
		}
	}

	private static NBTTagCompound getCompoundOrNull(ItemStack stack) {
		final NBTTagCompound compound = stack.getTagCompound();
		if (compound != null && compound.hasKey("BlockEntityTag")) {
			return compound.getCompoundTag("BlockEntityTag");
		} else {
			return null;
		}
	}
}

File: src/main/kotlin/nessiesson/usefulmod/ShulkerBoxDisplay.kt
package nessiesson.usefulmod

import nessiesson.usefulmod.config.Config
import net.minecraft.block.BlockShulkerBox
import net.minecraft.client.Minecraft
import net.minecraft.client.gui.Gui
import net.minecraft.client.gui.GuiScreen
import net.minecraft.client.renderer.GlStateManager
import net.minecraft.client.renderer.RenderHelper
import net.minecraft.client.resources.I18n
import net.minecraft.inventory.ItemStackHelper
import net.minecraft.item.ItemBlock
import net.minecraft.item.ItemShulkerBox
import net.minecraft.item.ItemStack
import net.minecraft.nbt.NBTTagCompound
import net.minecraft.util.NonNullList
import net.minecraft.util.ResourceLocation

// Modified version of the shulkerbox display from Vazkii's Quark Forgemod.
object ShulkerBoxDisplay {
	private val WIDGET_RESOURCE = ResourceLocation("usefulmod", "textures/shulker_widget.png")

	fun handleShulkerBoxDisplayRenderer(stack: ItemStack?, x: Int, y: Int, gui: Gui) {
		if (!Config.isShulkerBoxDisplayEnabled) {
			return
		}

		if (stack != null && stack.item is ItemShulkerBox && stack.hasTagCompound() && GuiScreen.isShiftKeyDown()) {
			val cmp = getCompoundOrNull(stack)
			if (cmp != null && cmp.hasKey("Items", 9)) {
				val texWidth = 172
				val texHeight = 64
				val currentY = y - texHeight - 18

				GlStateManager.pushMatrix()
				RenderHelper.enableGUIStandardItemLighting()
				GlStateManager.enableRescaleNormal()
				GlStateManager.translate(0f, 0f, 700f)

				val mc = Minecraft.getMinecraft()
				mc.textureManager.bindTexture(WIDGET_RESOURCE)

				val dye = ((stack.item as ItemBlock).block as BlockShulkerBox).color
				val colours = dye.colorComponentValues
				GlStateManager.color(colours[0], colours[1], colours[2])

				gui.drawTexturedModalRect(x, currentY, 0, 0, texWidth, texHeight)

				val itemList = NonNullList.withSize(27, ItemStack.EMPTY)
				ItemStackHelper.loadAllItems(cmp, itemList)

				val render = mc.renderItem
				GlStateManager.enableDepth()
				for ((i, item) in itemList.withIndex()) {
					if (!item.isEmpty) {
						val xp = x + 6 + i % 9 * 18
						val yp = currentY + 6 + i / 9 * 18
						render.renderItemAndEffectIntoGUI(item, xp, yp)
						render.renderItemOverlays(mc.fontRenderer, item, xp, yp)
					}
				}
				GlStateManager.disableDepth()
				GlStateManager.popMatrix()
			}
		}
	}

	fun addShulkerBoxTooltop(stack: ItemStack, tooltip: MutableList<String>) {
		if (!Config.isShulkerBoxDisplayEnabled) {
			return
		}

		if (!stack.isEmpty && stack.item is ItemShulkerBox && stack.hasTagCompound()) {
			val cmp = getCompoundOrNull(stack)
			if (cmp != null && cmp.hasKey("Items", 9)) {

				if (!GuiScreen.isShiftKeyDown()) {
					tooltip.add(1, I18n.format("usefulmod.ui.mention_shift"))
				}
			}
		}
	}

	private fun getCompoundOrNull(stack: ItemStack): NBTTagCompound? {
		val compound = stack.tagCompound
		return if (compound != null && compound.hasKey("BlockEntityTag")) {
			compound.getCompoundTag("BlockEntityTag")
		} else {
			null
		}
	}
}
